### PR TITLE
Collations 1.4

### PR DESCRIPTION
**Pull request by vmg**
## Description

This week's collations work revolves around two features:

- **Optimized codegen for contractions**: finding contractions in Unicode strings is one of the most expensive parts of comparing strings with UCA collations. The old implementation was using a trie to match on every position in the string trying to find the longest run of characters that forms an actual contraction -- this is similar to what MySQL was doing, but they have some internal optimizations which I cannot reverse engineer (remember GPLv2!) that makes their matching faster than ours.

    ...No more! I've come up with an extremely efficient way to find contractions. Since we were already generating contraction tables for the strings, it occured to me that we can generate optimized code that finds the contractions linearly without having to depend on a trie. The results are very good for all languages:

```
BenchmarkUCAContractions
BenchmarkUCAContractions/utf16_czech_ci-0.05-fast
BenchmarkUCAContractions/utf16_czech_ci-0.05-fast-16         	    4617	    252483 ns/op	 129.78 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_czech_ci-0.05-trie
BenchmarkUCAContractions/utf16_czech_ci-0.05-trie-16         	    2721	    421549 ns/op	  77.73 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_danish_ci-0.05-fast
BenchmarkUCAContractions/utf16_danish_ci-0.05-fast-16        	    4538	    260318 ns/op	 125.88 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_danish_ci-0.05-trie
BenchmarkUCAContractions/utf16_danish_ci-0.05-trie-16        	    2619	    445075 ns/op	  73.62 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_lithuanian_ci-0.05-fast
BenchmarkUCAContractions/utf16_lithuanian_ci-0.05-fast-16    	    4598	    252994 ns/op	 129.52 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_lithuanian_ci-0.05-trie
BenchmarkUCAContractions/utf16_lithuanian_ci-0.05-trie-16    	    2919	    425123 ns/op	  77.08 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_spanish2_ci-0.05-fast
BenchmarkUCAContractions/utf16_spanish2_ci-0.05-fast-16      	    3547	    335215 ns/op	  97.75 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_spanish2_ci-0.05-trie
BenchmarkUCAContractions/utf16_spanish2_ci-0.05-trie-16      	    2760	    447677 ns/op	  73.20 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_croatian_ci-0.05-fast
BenchmarkUCAContractions/utf16_croatian_ci-0.05-fast-16      	    3450	    337727 ns/op	  97.03 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf16_croatian_ci-0.05-trie
BenchmarkUCAContractions/utf16_croatian_ci-0.05-trie-16      	    2710	    437642 ns/op	  74.87 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf8mb4_cs_0900_ai_ci-0.05-fast
BenchmarkUCAContractions/utf8mb4_cs_0900_ai_ci-0.05-fast-16  	    4531	    258210 ns/op	 126.90 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf8mb4_cs_0900_ai_ci-0.05-trie
BenchmarkUCAContractions/utf8mb4_cs_0900_ai_ci-0.05-trie-16  	    2834	    430176 ns/op	  76.17 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf8mb4_da_0900_ai_ci-0.05-fast
BenchmarkUCAContractions/utf8mb4_da_0900_ai_ci-0.05-fast-16  	    4800	    258795 ns/op	 126.62 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf8mb4_da_0900_ai_ci-0.05-trie
BenchmarkUCAContractions/utf8mb4_da_0900_ai_ci-0.05-trie-16  	    2628	    443604 ns/op	  73.87 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf8mb4_lt_0900_ai_ci-0.05-fast
BenchmarkUCAContractions/utf8mb4_lt_0900_ai_ci-0.05-fast-16  	    5028	    231650 ns/op	 141.45 MB/s	       0 B/op	       0 allocs/op
BenchmarkUCAContractions/utf8mb4_lt_0900_ai_ci-0.05-trie
BenchmarkUCAContractions/utf8mb4_lt_0900_ai_ci-0.05-trie-16  	    2917	    399542 ns/op	  82.01 MB/s	       0 B/op	       0 allocs/op
````

- **Hash implementations**: we're using hashes already for DISTINCT in our query planner. In order for @king-11 to wire up the next part of the planner (after what he's currently working on), we need collation aware hashes. This PR implements them for all our supported collations. They're based on the mixing hash that ships with the Go runtime, so the actual hashing operation is quite fast in practice. Not much else to say about them, the implementation was straightforward.

## Related Issue(s)
- https://github.com/vitessio/vitess/issues/8606

## Checklist
- [x] Tests were added or are not required
- [x] Documentation was added or is not required

### CODE REVIEW

File: go/mysql/collations/tools/makemysqldata/contractions.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
+package main
```
Comment (systay) [inline go/mysql/collations/tools/makemysqldata/contractions.go:1]: license header missing

File: go/mysql/collations/hash.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	"unsafe"
+)
+
+const (
+	m1 = 0xa0761d6478bd642f
+	m2 = 0xe7037ed1a0b428db
+	m3 = 0x8ebc6af09c88c6e3
+	m4 = 0x589965cc75374cc3
+	m5 = 0x1d8e4e27c47d124f
+)
```
Comment (frouioui) [inline go/mysql/collations/hash.go:30]: What are these? Arbitrary hashes?

File: go/mysql/collations/hash.go [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+func memhash(src []byte, seed uintptr) uintptr {
+	srclen := len(src)
+	if srclen == 0 {
+		return seed ^ m1
+	}
+	return memhashraw(unsafe.Pointer(&src[0]), seed, uintptr(srclen))
+}
+
+func memhashraw(p unsafe.Pointer, seed, s uintptr) uintptr {
+	var a, b uintptr
```
Comment (frouioui) [inline go/mysql/collations/hash.go:41]: this function looks like pure magic 🧙🏻‍♂️

File: go/mysql/collations/uca_contraction_test.go [hunk 4]
```diff
@@ -0,0 +1,254 @@
+package collations
+
```
Comment (frouioui) [inline go/mysql/collations/uca_contraction_test.go:2]: header is missing

**Review by frouioui — APPROVED** [review]
looks good to me 👍🏻

File: go/mysql/collations/hash.go (hunk 2 above)
Comment (vmg) [reply go/mysql/collations/hash.go:30]: Yes, these are the same hash constants that are used by the Go runtime:

https://golang.org/src/runtime/hash64.go

**Pull request merged**